### ISSUE TITLE
Add optional parameter on functions like Elements.Button().primary()

### ISSUE THREAD

**Issue by chriscmuir** [raycharius/slack-block-builder#30]
Hi Ray

For functions like Elements.Button().primary() which are ultimately just setting a boolean flag for component, can we please add an optional boolean parameter to these functions that allows us to pass an expression?

This would allow me to do something like this, where one of 3 buttons could be primary based on an expression:

    ui.Blocks.Actions()
      .elements(
        ui.Elements.Button()
        .text('Books')
        .value('books')
        .primary(properties.currentOption === 'books'),
      ui.Elements.Button()
        .text('Magazines')
        .value('magazines')
        .primary(properties.currentOption === 'magazines'),
      ui.Elements.Button()
        .text('News')
        .value('news')
        .primary(properties.currentOption === 'news'),

Otherwise I have to write a bunch of if statements and push each element to the ui.Blocks.Actions().elements() method, which makes for clunky syntax.

Thanks,

Chris

**Comment by raycharius** [comment]
Thanks for the feedback! Overall, does sound like a good idea. Let me think this over and try to scope it out. This is possible with the appending methods, so it does make sense to have that available for these methods, too.

**Comment by raycharius** [comment]
@chriscmuir What do you think about a separate `style()` method?

Something like:

``` javascript
Blocks.Actions()
  .elements(
    Elements.Button()
      .text('Books')
      .value('books')
      .style(properties.currentOption === 'books' ? 'primary' : null),
```

It may seem a bit clunky here, but it leaves more opportunities if more styles were to be introduced.

**Comment by chriscmuir** [comment]
Thanks for following up @raycharius .

I'm of two minds on your proposal.  I appreciate the point that if more styles are introduced, you're proposed solution would make it easier for you to support them.  You would have to document the style properties of course, but I guess they would align with the block kit equivalents and is easily guessable by new programmers picking up Block Builder.

Conversely the Block Builder code library currently provides functions for the block kit UI elements, such as the .primary() function.  To start providing a dynamic style() function vs a function based solution seems to go against what's been built so far.

Just my $0.02 worth though, and in reality either solution would work.  Thanks again for looking into this, and also providing the Block Builder library, I appreciate the work you've put into this.

**Comment by raycharius** [comment]
Yes, very good point on it moving away from the conventions already introduced. I guess the biggest question in my head here is – can this be behavior on the `Builder.set()` method.

I'm currently working on v2.0.0, rewriting in TS, restructuring the code to make changes like this easier, and adding a new `Blocks` container, as well a few components in the works (`Paginator` and `Accordion`), and I've added this to that milestone.

I'm hoping to get that out in the next month or so. But if you want, feel free to send over a PR in the meantime, adding this locally to the `danger()` and `primary()` methods, if you need this earlier – happy to take a look at how this could be solved!

I'm also hoping to get some contributing guidelines up in the next week or so, but the only one thing I try to be adamant about is [emoji commit messages](https://gist.github.com/parmentf/035de27d6ed1dce0b36a). And 100% test coverage!

**Comment by raycharius** [comment]
Currently working on v2.0.0, and what I've done is make it so that all of the 'configuration' methods that typically do not take an argument (`danger()`, `primary()`, `dispatchAction()`, `ignoreMarkdown()`, etc.), have a default value, but can also be passed a boolean, allowing for conditionals. Expect to release this within in the week, 2.0.0b

``` typescript
export abstract class HasDanger extends PropSetter {
  public danger(boolean = true): this {
    if (boolean) {
      return this.set(values.danger, props.style);
    }

    return this;
  }
}

```

``` typescript
export abstract class HasDispatchAction extends PropSetter {

  public dispatchAction(boolean = true): this {
    return this.set(boolean, props.dispatchAction);
  }
}
```

**Comment by raycharius** [comment]
Enjoy! It is currently available via the v2.0.0 Beta.

``` bash
npm install --save slack-block-builder@2.0.0-beta.1
```
